contracts/governance/extensions/GovernorSequentialProposalId.sol

```diff
@@ -0,0 +1,32 @@
+// SPDX-License-Identifier: MIT
+
+pragma solidity ^0.8.20;
+
+import {Governor} from "../Governor.sol";
+
+abstract contract GovernorSequentialProposalId is Governor {
+    error NextProposalIdCanOnlyBeSetOnce();
+
+    uint256 private _numberOfProposals;
+    mapping(uint256 proposalHash => uint256 proposalId) private _proposalIds;
+
+    function _getProposalId(
+        address[] memory targets,
+        uint256[] memory values,
+        bytes[] memory calldatas,
+        bytes32 descriptionHash
+    ) internal virtual override returns (uint256) {
+        uint256 proposalHash = super._getProposalId(targets, values, calldatas, descriptionHash);
+
+        uint256 storedProposalId = _proposalIds[proposalHash];
+        return storedProposalId == 0 ? (_proposalIds[proposalHash] = ++_numberOfProposals) : storedProposalId;
+    }
+
+    /**
+     * @dev Internal function to set the sequential proposal ID for the next proposal. This is helpful for transitioning from another governing system.
+     */
+    function _setNextProposalId(uint256 nextProposalId) internal virtual {
+        if (_numberOfProposals != 0) revert NextProposalIdCanOnlyBeSetOnce();
```

Comment:
IMO this works, but there's a slightly annoying attack surface here. if an attacker can create a proposal on this governor before the upgrade occurs, then any `setNextProposalId` call will revert. this `setNextProposalId` call might be part of the upgrade proposal, which would then fail.

we had the same problem. [we managed it](https://github.com/ScopeLift/compound-governance-upgrade/pull/43/files#diff-8dfda4c5f188c565ed922bb6417549df160fa6898ffd3b316c2b8ea8e888c4abR66-R72) by initializing `nextProposalId` to `type(uint256).max`. then, proposals couldn't be created unless `setNextProposalId` had been called.

Reply:
Given that the default flow would be proposal ids starting from 1, we don't want to require calls to `_setNextProposalId`. Also note that `_setNextProposalId` is an internal function which would need to be called in a bespoke public function which is where this issue should be addressed. We should add another function to this extension `getNextProposalId` which would make disabling governance until the proposal id is set easier.